sonic-net/sonic-buildimage
# Fix the bug related to stack overflow in iccpd.

iccpd: fix a bug related to stack overflow in iccpd(update_peerlink_isolate_from_all_csm_lif) when the number of mclag members > 32 or more

* change array length for 'mlag_po_buf' to 2048.
* Add length check before using 'snprintf'. Note:The return value of the snprintf function is the length of the source string.

  Signed-off-by: ccyyrr92 cuiyingru@asterfusion.com

#### Why I did it
The return value of the snprintf function is the length of the source string.
In the following code: if 'the length of the source string' > 'the length of the dst string', snprintf will also return 'the length of the source string'. then the dst_len may be greater than sizeof(mlag_po_buf), then 'sizeof(mlag_po_buf) - dst_len' may be a negative. then lead to a stack overflow!
----------------------------------------------------------------------------------------------------------------
    if (strlen(mlag_po_buf) != 0)
    {
        dst_len += snprintf(mlag_po_buf + dst_len, sizeof(mlag_po_buf) - dst_len, "%s", ",");
    }

    dst_len += snprintf(mlag_po_buf + dst_len, sizeof(mlag_po_buf) - dst_len, "%s%s%s",
                         lif->name, lif->portchannel_member_buf[0] == 0 ? "" : ",", lif->portchannel_member_buf);
----------------------------------------------------------------------------------------------------------------

##### Work item tracking
no

#### How I did it
I add the length of array 'mlag_po_buf'. 
I add the length checking when using snprintf.
#### How to verify it
when I test, I change the length of array 'mlag_po_buf' to 20.
then config:
----------------------------------------
vlan 100
vlan 200
vlan 300
interface link-aggregation 1
 switchport access vlan 100
interface link-aggregation 2
 switchport access vlan 300
interface ethernet 5
 link-aggregation-group 1
interface ethernet 9
 link-aggregation-group 2
interface ethernet 17
 speed 1000
 switchport trunk vlan 100
 switchport trunk vlan 200
 switchport trunk vlan 300
interface vlan 200
 ip address 10.1.1.48/24
mclag domain 1
 local-address 10.1.1.48
 peer-address 10.1.1.24
 peer-link ethernet 17
------------------------------------------------------------
Confirm that the MCLAG connection has been successfully established.
then config:
-------------------------------------------
mclag domain 1
 member lag 1
 member lag 2
--------------------------------------------------
then I show the warning log . and there is no stack overflow.



Please help me review the modify for the question.Thanks
